File: php/math/collatz.php
<?php
// Print the start below 1000 with the longest Collatz sequence, tracked in a running maximum.

declare(strict_types=1);

const LIMIT = 1000;

function chainLength(int $start): int
{
    $length = 1;
    while ($start !== 1) {
        $start = $start % 2 === 0 ? intdiv($start, 2) : $start * 3 + 1;
        $length++;
    }

    return $length;
}

$longest = 1;
$best = 1;

for ($start = 1; $start < LIMIT; $start++) {
    $length = chainLength($start);
    if ($length > $best) {
        $longest = $start;
        $best = $length;
    }
}

echo "$longest $best\n";

File: clojure/math/collatz.clj
;; Print the start below 1000 with the longest Collatz sequence, picked by max-key over a range.
;; Run: clojure -M collatz.clj

(defn step [n]
  (if (even? n) (quot n 2) (inc (* n 3))))

(defn chain-length [start]
  (inc (count (take-while #(not= 1 %) (iterate step start)))))

(let [longest (apply max-key chain-length (range 1 1000))]
  (println longest (chain-length longest)))
